### PR TITLE
[cominterop] Always marshal a class as a COM interface if it has the InterfaceType attribute set.

### PR DESCRIPTION
**Pull request by zfigura**
This fixes Game Maker Studio 2 Desktop under Wine.

### CODE REVIEW

**Comment by marek-safar** [comment]
approve

**Comment by marek-safar** [comment]
@vargaz @lambdageek could you please review

File: mono/metadata/cominterop.c [hunk 1]
```diff
@@ -3756,3 +3756,26 @@ mono_cominterop_get_com_interface_internal (gboolean icall, MonoObjectHandle obj
...
+#ifndef DISABLE_COM
+	ERROR_DECL (error);
+	MonoCustomAttrInfo* cinfo = NULL;
+	MonoInterfaceTypeAttribute* itf_attr = NULL;
+
+	cinfo = mono_custom_attrs_from_class_checked (klass, error);
+	mono_error_assert_ok (error);
+	if (cinfo) {
+		itf_attr = (MonoInterfaceTypeAttribute*)mono_custom_attrs_get_attr_checked (cinfo, mono_class_get_interface_type_attribute_class (), error);
+		mono_error_assert_ok (error); /*FIXME proper error handling*/
```
Comment (lambdageek) [inline mono/metadata/cominterop.c]: What do we want proper error handling to be?  Can we just return false if trying to look for the custom attribute fails?  In that case could do something like this:

```c
if (!is_ok (error)) {
    mono_error_cleanup (error);
    itf_attr = NULL;
}
```

Comment (zfigura) [reply mono/metadata/cominterop.c]: I essentially just copy-pasted that part from cominterop_get_com_slot_begin (admittedly, I should have at least separated it out into a common helper). I wasn't really sure what proper error handling would be, but that seems reasonable enough. I'll resend the patch.

**Comment by lambdageek** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins build

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by akoeplinger** [comment]
This seems to cause two test failures, one during `aot-test` when trying to AOT System.Data.dll and one during `fullaot-mixed` which crashes with this error when AOT-ing System.dll:

```
Invoking method 'System.Runtime.InteropServices.InterfaceTypeAttribute:.ctor (System.Runtime.InteropServices.ComInterfaceType)' when running in no-exec mode.
```

**Comment by zfigura** [comment]
Indeed, thanks. In `mono_marshal_get_native_wrapper()` I see a comment "The COM code is not AOT compatible, it calls mono_custom_attrs_get_attr_checked ()". Is there by any chance a simple explanation for why this is the case?

**Comment by zfigura** [comment]
Okay, thanks to some hints from Zoltan Varga, I believe I've figured out the right way to resolve this. I've pushed a new version that avoids using `mono_custom_attrs_get_attr_checked()`.

**Pull request merged**